File: src/fim_one/web/schemas/workflow.py
"""Workflow request/response schemas."""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, Field, field_validator


# ---------------------------------------------------------------------------
# Create / Update
# ---------------------------------------------------------------------------


class WorkflowCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    icon: str | None = None
    description: str | None = None
    blueprint: dict = Field(
        default_factory=lambda: {"nodes": [], "edges": [], "viewport": {}}
    )
    status: str = "draft"
    is_active: bool = True


class WorkflowUpdate(BaseModel):
    name: str | None = None
    icon: str | None = None
    description: str | None = None
    blueprint: dict | None = None
    status: str | None = None
    is_active: bool | None = None
    webhook_url: str | None = None


# ---------------------------------------------------------------------------
# Response
# ---------------------------------------------------------------------------


class WorkflowResponse(BaseModel):
    id: str
    user_id: str
    name: str
    icon: str | None
    description: str | None
    blueprint: dict
    input_schema: dict | None
    output_schema: dict | None
    status: str
    is_active: bool = True
    visibility: str = "personal"
    org_id: str | None = None
    publish_status: str | None = None
    published_at: str | None = None
    reviewed_by: str | None = None
    reviewed_at: str | None = None
    review_note: str | None = None
    webhook_url: str | None = None
    schedule_cron: str | None = None
    schedule_enabled: bool = False
    schedule_inputs: dict[str, Any] | None = None
    schedule_timezone: str = "UTC"
    has_api_key: bool = False
    total_runs: int = 0
    last_run_at: str | None = None
    success_rate: float | None = None
    created_at: str
    updated_at: str | None


# ---------------------------------------------------------------------------
# Run
# ---------------------------------------------------------------------------


class WorkflowRunRequest(BaseModel):
    """Input payload to execute a workflow."""

    inputs: dict[str, Any] = Field(default_factory=dict)
    dry_run: bool = Field(
        default=False,
        description="When true, validate and return the execution plan without running.",
    )


class NodeRunResult(BaseModel):
    """Result of a single node execution."""

    node_id: str
    node_type: str
    status: str  # completed | failed | skipped
    output: Any = None
    error: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    duration_ms: int | None = None


class WorkflowRunResponse(BaseModel):
    id: str
    workflow_id: str
    user_id: str
    status: str
    inputs: dict | None
    outputs: dict | None
    node_results: dict | None
    started_at: str | None
    completed_at: str | None
    duration_ms: int | None
    error: str | None
    created_at: str
    updated_at: str | None


# ---------------------------------------------------------------------------
# Batch Run
# ---------------------------------------------------------------------------


class WorkflowBatchRunRequest(BaseModel):
    """Input payload to execute a workflow with multiple input sets."""

    inputs: list[dict[str, Any]] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="List of input sets to run the workflow with (1-100 items).",
    )
    max_parallel: int = Field(
        default=3,
        ge=1,
        le=10,
        description="Maximum number of concurrent workflow executions (1-10).",
    )


class BatchRunResultItem(BaseModel):
    """Result of a single workflow execution within a batch."""

    run_id: str
    inputs: dict[str, Any]
    status: str  # completed | failed | cancelled
    outputs: dict[str, Any] | None = None
    error: str | None = None
    duration_ms: int | None = None


class WorkflowBatchRunResponse(BaseModel):
    """Response for a batch workflow execution."""

    batch_id: str
    total: int
    results: list[BatchRunResultItem] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Validate / Dry Run
# ---------------------------------------------------------------------------


class BlueprintWarningItem(BaseModel):
    """A single non-fatal validation warning."""

    node_id: str | None = None
    code: str
    message: str


class WorkflowValidateResponse(BaseModel):
    """Structural validation result for a workflow blueprint."""

    valid: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[BlueprintWarningItem] = Field(default_factory=list)
    node_count: int = 0
    edge_count: int = 0
    topology_order: list[str] = Field(default_factory=list)


class DryRunNodePlan(BaseModel):
    """Planned execution info for a single node in dry-run mode."""

    node_id: str
    node_type: str
    position: int  # 0-based index in the execution order
    has_warnings: bool = False


class WorkflowDryRunResponse(BaseModel):
    """Dry-run result: execution plan without actual execution."""

    valid: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[BlueprintWarningItem] = Field(default_factory=list)
    node_count: int = 0
    edge_count: int = 0
    topology_order: list[str] = Field(default_factory=list)
    execution_plan: list[DryRunNodePlan] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Version
# ---------------------------------------------------------------------------


class WorkflowVersionResponse(BaseModel):
    id: str
    workflow_id: str
    version_number: int
    blueprint: dict
    input_schema: dict | None
    output_schema: dict | None
    change_summary: str | None
    created_by: str | None
    created_at: str


# ---------------------------------------------------------------------------
# Export / Import
# ---------------------------------------------------------------------------


class WorkflowExportData(BaseModel):
    """Portable workflow representation for export (inner payload)."""

    name: str
    icon: str | None = None
    description: str | None = None
    blueprint: dict
    input_schema: dict | None = None
    output_schema: dict | None = None


class WorkflowExportFile(BaseModel):
    """Top-level envelope for exported workflow files."""

    format: str = "fim_workflow_v1"
    exported_at: str
    workflow: WorkflowExportData


class WorkflowImportRequest(BaseModel):
    """Request body for importing a workflow (legacy wrapper)."""

    data: WorkflowExportData


class WorkflowImportFileRequest(BaseModel):
    """Request body matching the exported file envelope.

    Accepts ``{ "format": "fim_workflow_v1", "exported_at": ..., "workflow": {...} }``
    as well as the legacy ``{ "data": {...} }`` shape.
    """

    format: str | None = None
    exported_at: str | None = None
    workflow: WorkflowExportData | None = None
    # Legacy fallback
    data: WorkflowExportData | None = None


class UnresolvedReferenceItem(BaseModel):
    """A single unresolved external resource reference in an imported blueprint."""

    node_id: str
    node_type: str
    field_name: str
    referenced_id: str
    resource_type: str


class WorkflowImportResponse(BaseModel):
    """Response for workflow import — includes the created workflow plus any
    unresolved reference warnings."""

    workflow: WorkflowResponse
    unresolved_references: list[UnresolvedReferenceItem] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Env vars
# ---------------------------------------------------------------------------


class WorkflowEnvVarsUpdate(BaseModel):
    """Encrypted env vars key-value pairs."""

    env_vars: dict[str, str] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Duplicate / Templates
# ---------------------------------------------------------------------------


class WorkflowFromTemplateRequest(BaseModel):
    """Request body for creating a workflow from a template (built-in or DB-stored)."""

    template_id: str = Field(min_length=1)
    name: str | None = None


class WorkflowTemplateResponse(BaseModel):
    """A workflow template descriptor (built-in or DB-stored)."""

    id: str
    name: str
    description: str
    icon: str
    category: str
    blueprint: dict
    created_at: str | None = None


class WorkflowTemplateCreate(BaseModel):
    """Request body for creating a DB-stored workflow template (admin only)."""

    name: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1)
    icon: str = "🔄"
    category: str = Field(min_length=1, max_length=100)
    blueprint: dict
    is_active: bool = True
    sort_order: int = 0


class WorkflowTemplateUpdate(BaseModel):
    """Request body for updating a DB-stored workflow template (admin only)."""

    name: str | None = None
    description: str | None = None
    icon: str | None = None
    category: str | None = None
    blueprint: dict | None = None
    is_active: bool | None = None
    sort_order: int | None = None


# ---------------------------------------------------------------------------
# Analytics
# ---------------------------------------------------------------------------


class RunsPerDay(BaseModel):
    """Aggregated run counts for a single calendar day."""

    date: str
    count: int
    completed: int = 0
    failed: int = 0


class MostFailedNode(BaseModel):
    """A node that has failed across workflow runs."""

    node_id: str
    failure_count: int
    total_runs: int


class WorkflowAnalyticsResponse(BaseModel):
    """Comprehensive workflow execution analytics."""

    total_runs: int
    status_distribution: dict[str, int]
    success_rate: float | None = None
    avg_duration_ms: int | None = None
    p50_duration_ms: int | None = None
    p95_duration_ms: int | None = None
    p99_duration_ms: int | None = None
    runs_per_day: list[RunsPerDay] = Field(default_factory=list)
    most_failed_nodes: list[MostFailedNode] = Field(default_factory=list)
    avg_nodes_per_run: float | None = None


# ---------------------------------------------------------------------------
# Schedule (cron-based trigger)
# ---------------------------------------------------------------------------

# Allowed ranges for 5-field cron expressions (min hour dom month dow)
_CRON_FIELD_RANGES = [
    (0, 59),   # minute
    (0, 23),   # hour
    (1, 31),   # day of month
    (1, 12),   # month
    (0, 7),    # day of week (0 and 7 both = Sunday)
]

# Matches a single cron token: number, range (1-5), step (*/2, 1-5/2), or list (1,3,5)
# Also allows named days/months (MON-FRI, JAN-DEC)
_CRON_TOKEN_RE = re.compile(
    r"^(?:\*|[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)?)(?:/\d+)?$"
)


def _validate_cron(expr: str) -> str | None:
    """Validate a 5-field cron expression.

    Returns None if valid, or an error message string if invalid.
    Supports: ``*``, ranges (``1-5``), steps (``*/2``), lists (``1,3,5``),
    and named days/months (``MON-FRI``, ``JAN``).
    """
    parts = expr.strip().split()
    if len(parts) != 5:
        return f"Cron expression must have exactly 5 fields, got {len(parts)}"

    for i, part in enumerate(parts):
        # A field can be a comma-separated list of tokens
        tokens = part.split(",")
        for token in tokens:
            if not token:
                return f"Empty token in field {i + 1}"
            if not _CRON_TOKEN_RE.match(token):
                return f"Invalid token '{token}' in field {i + 1}"

    return None


def _compute_next_run(cron_expr: str, tz_name: str = "UTC") -> str | None:
    """Compute the next scheduled run time from a cron expression.

    Uses ``croniter`` if available; otherwise returns ``None``.
    """
    try:
        from croniter import croniter
    except ImportError:
        return None

    try:
        import zoneinfo

        tz = zoneinfo.ZoneInfo(tz_name)
    except Exception:
        tz = timezone.utc

    now = datetime.now(tz=tz)
    try:
        it = croniter(cron_expr, now)
        next_dt: datetime = it.get_next(datetime)
        return next_dt.isoformat()
    except Exception:
        return None


class WorkflowScheduleUpdate(BaseModel):
    """Request body to set or update a workflow's scheduled trigger."""

    cron: str | None = Field(
        None,
        max_length=100,
        description="5-field cron expression (min hour dom month dow).",
        examples=["0 9 * * MON-FRI", "*/15 * * * *"],
    )
    enabled: bool = False
    inputs: dict[str, Any] | None = None
    timezone: str = Field(
        default="UTC",
        max_length=50,
        description="IANA timezone name for schedule evaluation.",
    )

    @field_validator("cron")
    @classmethod
    def validate_cron_format(cls, v: str | None) -> str | None:
        if v is None:
            return v
        err = _validate_cron(v)
        if err:
            raise ValueError(err)
        return v.strip()


class WorkflowScheduleResponse(BaseModel):
    """Current schedule configuration for a workflow."""

    schedule_cron: str | None = None
    schedule_enabled: bool = False
    schedule_inputs: dict[str, Any] | None = None
    schedule_timezone: str = "UTC"
    next_run_at: str | None = None


# ---------------------------------------------------------------------------
# Public trigger (API key)
# ---------------------------------------------------------------------------


class WorkflowTriggerRequest(BaseModel):
    """Input payload for the public trigger endpoint."""

    inputs: dict[str, Any] = Field(default_factory=dict)


class WorkflowTriggerResponse(BaseModel):
    """Response from the public trigger endpoint."""

    run_id: str
    status: str
    outputs: dict[str, Any] | None = None
    error: str | None = None
    duration_ms: int | None = None


class WorkflowApiKeyResponse(BaseModel):
    """Response after generating an API key (shown once)."""

    api_key: str
    workflow_id: str


# ---------------------------------------------------------------------------
# Test Node (single-node isolated execution)
# ---------------------------------------------------------------------------


class NodeTestRequest(BaseModel):
    """Request body for testing a single workflow node in isolation."""

    node_id: str = Field(min_length=1, description="ID of the node to test")
    variables: dict[str, Any] = Field(
        default_factory=dict,
        description="Mock variable values to populate the variable store",
    )
    env_vars: dict[str, str] = Field(
        default_factory=dict,
        description="Optional environment variable overrides for this test run",
    )


class NodeTestResponse(BaseModel):
    """Result of a single-node test execution."""

    node_id: str
    node_type: str
    status: str  # completed | failed
    output: Any = None
    error: str | None = None
    duration_ms: int
    variables_after: dict[str, Any] = Field(
        default_factory=dict,
        description="Variable store snapshot after execution",
    )


# ---------------------------------------------------------------------------
# Admin batch operations
# ---------------------------------------------------------------------------


class BatchWorkflowDeleteRequest(BaseModel):
    """Request body for batch-deleting workflows."""

    workflow_ids: list[str] = Field(..., min_length=1, max_length=100)


class BatchWorkflowToggleRequest(BaseModel):
    """Request body for batch-toggling workflow is_active state."""

    workflow_ids: list[str] = Field(..., min_length=1, max_length=100)
    is_active: bool


class BatchWorkflowPublishRequest(BaseModel):
    """Request body for batch-publishing/unpublishing workflows."""

    workflow_ids: list[str] = Field(..., min_length=1, max_length=100)
    status: str  # "active" or "draft"


class BatchOperationResponse(BaseModel):
    """Response for batch operations."""

    count: int
    message: str

File: frontend/src/types/workflow.ts
// --- Core workflow types ---

export interface WorkflowResponse {
  id: string
  user_id: string
  name: string
  icon: string | null
  description: string | null
  blueprint: WorkflowBlueprint
  input_schema: Record<string, unknown> | null
  output_schema: Record<string, unknown> | null
  status: "draft" | "active"
  is_active: boolean
  visibility: string
  org_id?: string | null
  publish_status: string | null
  published_at: string | null
  reviewed_by: string | null
  reviewed_at: string | null
  review_note: string | null
  webhook_url: string | null
  schedule_cron?: string | null
  schedule_enabled?: boolean
  schedule_inputs?: Record<string, unknown> | null
  schedule_timezone?: string
  has_api_key: boolean
  forked_from: string | null
  total_runs: number
  last_run_at: string | null
  success_rate: number | null
  created_at: string
  updated_at: string
}

export interface WorkflowVariable {
  name: string
  type: "string" | "number" | "boolean" | "json"
  default_value: string
  description: string
}

export interface WorkflowBlueprint {
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
  viewport: { x: number; y: number; zoom: number }
  variables?: WorkflowVariable[]
}

export type ErrorStrategy = "stop_workflow" | "continue" | "fail_branch"

export interface WorkflowNode {
  id: string
  type: WorkflowNodeType
  position: { x: number; y: number }
  data: Record<string, unknown>
  error_strategy?: ErrorStrategy
  timeout_ms?: number
}

export interface WorkflowEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string
  targetHandle?: string
}

export type WorkflowNodeType =
  | "start"
  | "end"
  | "llm"
  | "conditionBranch"
  | "questionClassifier"
  | "agent"
  | "knowledgeRetrieval"
  | "connector"
  | "httpRequest"
  | "variableAssign"
  | "templateTransform"
  | "codeExecution"
  | "iterator"
  | "loop"
  | "variableAggregator"
  | "parameterExtractor"
  | "listOperation"
  | "transform"
  | "documentExtractor"
  | "questionUnderstanding"
  | "humanIntervention"
  | "mcp"
  | "builtinTool"
  | "subWorkflow"
  | "env"

// --- Per-node data interfaces ---

export interface StartNodeData {
  variables: Array<{
    name: string
    type: string
    default_value?: string
    required?: boolean
  }>
}

export interface EndNodeData {
  output_mapping: Record<string, string>
}

export interface LLMNodeData {
  model?: string
  model_tier?: "fast" | "main"
  system_prompt?: string
  prompt_template: string
  output_variable: string
  temperature?: number
  max_tokens?: number
}

export interface ConditionNodeData {
  mode: "expression" | "llm"
  conditions: Array<{
    id: string
    label: string
    expression?: string
    llm_prompt?: string
  }>
}

export interface QuestionClassifierNodeData {
  model?: string
  prompt?: string
  classes: Array<{
    id: string
    label: string
    description?: string
  }>
}

export interface AgentNodeData {
  agent_id: string
  prompt_template?: string
  output_variable: string
}

export interface KnowledgeRetrievalNodeData {
  kb_id: string
  query_template: string
  top_k?: number
  output_variable: string
}

export interface ConnectorNodeData {
  connector_id: string
  action: string
  parameters: Record<string, string>
  output_variable: string
}

export interface HTTPRequestNodeData {
  method: string
  url: string
  headers?: Record<string, string>
  body?: string
  output_variable: string
}

export interface VariableAssignNodeData {
  assignments: Array<{
    variable: string
    expression: string
  }>
}

export interface TemplateTransformNodeData {
  template: string
  output_variable: string
}

export interface CodeExecutionNodeData {
  language: "python" | "javascript"
  code: string
  output_variable: string
}

export interface IteratorNodeData {
  list_variable: string
  iterator_variable: string
  index_variable: string
  max_iterations: number
}

export interface LoopNodeData {
  condition: string
  max_iterations: number
  loop_variable: string
}

export interface VariableAggregatorNodeData {
  variables: string[]
  mode: "list" | "concat" | "merge" | "first_non_empty"
  separator: string
}

export interface ListOperationNodeData {
  input_variable: string
  operation: "filter" | "map" | "sort" | "slice" | "flatten" | "unique" | "reverse" | "length"
  expression: string
  slice_start?: number
  slice_end?: number
  output_variable: string
}

export interface TransformNodeData {
  input_variable: string
  operations: Array<{
    type: "json_path" | "type_cast" | "format" | "regex_extract" | "string_op" | "math_op"
    config: Record<string, unknown>
  }>
  output_variable: string
}

export interface DocumentExtractorNodeData {
  input_variable: string
  input_type: "text" | "base64" | "url"
  extract_mode: "full_text" | "pages" | "metadata" | "tables"
  page_range?: string
  output_variable: string
}

export interface QuestionUnderstandingNodeData {
  input_variable: string
  mode: "rewrite" | "expand" | "classify" | "decompose"
  system_prompt?: string
  output_variable: string
}

export interface HumanInterventionNodeData {
  prompt_message: string
  assignee: string
  timeout_hours: number
  output_variable: string
}

export interface MCPNodeData {
  server_id: string
  tool_name: string
  parameters: Record<string, unknown>
  output_variable: string
}

export interface BuiltinToolNodeData {
  tool_id: string
  parameters: Record<string, unknown>
  output_variable: string
}

export interface ParameterExtractorNodeData {
  input_text: string
  parameters: Array<{
    name: string
    type: string
    description: string
    required?: boolean
  }>
  extraction_prompt?: string
}

export interface SubWorkflowNodeData {
  workflow_id: string
  input_mapping: Record<string, string>
  output_variable: string
}

export interface ENVNodeData {
  env_keys: string[]
  output_variable: string
}


// --- Validation types ---

export interface BlueprintWarningItem {
  node_id: string | null
  code: string
  message: string
}

export interface WorkflowValidateResponse {
  valid: boolean
  errors: string[]
  warnings: BlueprintWarningItem[]
  node_count: number
  edge_count: number
  topology_order: string[]
}

// --- Import result types ---

export interface UnresolvedReference {
  node_id: string
  node_type: string
  field_name: string
  referenced_id: string
  resource_type: string
}

export interface WorkflowImportResult {
  workflow: WorkflowResponse
  unresolved_references: UnresolvedReference[]
  warnings: string[]
}

// --- Create / Update payloads ---

export interface WorkflowCreate {
  name: string
  icon?: string | null
  description?: string | null
  blueprint?: WorkflowBlueprint
}

export interface WorkflowUpdate {
  name?: string
  icon?: string | null
  description?: string | null
  blueprint?: WorkflowBlueprint
  input_schema?: Record<string, unknown> | null
  output_schema?: Record<string, unknown> | null
  status?: "draft" | "active"
  webhook_url?: string | null
}

// --- Run types ---

export interface WorkflowRunResponse {
  id: string
  workflow_id: string
  status: "pending" | "running" | "completed" | "failed" | "cancelled"
  inputs: Record<string, unknown> | null
  outputs: Record<string, unknown> | null
  node_results: Record<string, NodeRunResult> | null
  started_at: string | null
  completed_at: string | null
  duration_ms: number | null
  error: string | null
  created_at: string
}

export interface NodeRunResult {
  node_id?: string
  node_type?: string
  status: "pending" | "running" | "completed" | "failed" | "skipped" | "retrying"
  output: unknown
  error: string | null
  started_at: string | null
  completed_at: string | null
  duration_ms: number | null
  input_preview?: unknown
  retryAttempt?: number
  maxRetries?: number
  trace?: {
    variable_snapshot?: Record<string, unknown> | null
    input_snapshot?: Record<string, unknown> | null
    [key: string]: unknown
  }
}

/** Overlay data injected into node.data during workflow runs for canvas display */
export interface NodeRunOverlayData {
  durationMs: number | null
  inputPreview: string | null
  outputPreview: string | null
  runError: string | null
}

// --- Node run status for canvas overlay ---

export type NodeRunStatus = "pending" | "running" | "completed" | "failed" | "skipped" | "retrying"

// --- Execution log event for run panel ---

export type WorkflowLogEventType =
  | "node_started"
  | "node_completed"
  | "node_failed"
  | "node_skipped"
  | "node_retrying"
  | "run_completed"
  | "run_failed"

export interface WorkflowLogEvent {
  timestamp: number
  eventType: WorkflowLogEventType
  nodeId: string | null
  details: Record<string, unknown>
}

// --- Analytics (detailed) ---

export interface RunsPerDayEntry {
  date: string
  count: number
  completed: number
  failed: number
}

export interface MostFailedNodeEntry {
  node_id: string
  failure_count: number
  total_runs: number
}

export interface WorkflowAnalyticsResponse {
  total_runs: number
  status_distribution: Record<string, number>
  success_rate: number
  avg_duration_ms: number
  p50_duration_ms: number
  p95_duration_ms: number
  p99_duration_ms: number
  runs_per_day: RunsPerDayEntry[]
  most_failed_nodes: MostFailedNodeEntry[]
  avg_nodes_per_run: number
}

// --- Stats ---

export interface WorkflowStats {
  total_runs: number
  completed: number
  failed: number
  cancelled: number
  success_rate: number | null
  avg_duration_ms: number | null
  last_run_at: string | null
}

// --- Per-Node Stats (from /node-stats endpoint) ---

export interface NodeStatEntry {
  node_id: string
  total_runs: number
  completed: number
  failed: number
  skipped: number
  avg_duration_ms: number | null
  min_duration_ms: number | null
  max_duration_ms: number | null
  success_rate: number | null
}

export interface NodeStatsResponse {
  runs_analyzed: number
  nodes: NodeStatEntry[]
}

// --- Templates ---

export interface WorkflowTemplate {
  id: string
  name: string
  description: string
  icon: string
  category: string
  blueprint: WorkflowBlueprint
}

export interface WorkflowFromTemplateRequest {
  template_id: string
  name?: string
}

// --- Batch Run types ---

export interface BatchRunResultItem {
  run_id: string
  inputs: Record<string, unknown>
  status: "completed" | "failed"
  outputs: Record<string, unknown> | null
  error: string | null
  duration_ms: number | null
}

export interface WorkflowBatchRunResponse {
  batch_id: string
  total: number
  results: BatchRunResultItem[]
}

// --- Schedule types ---

export interface WorkflowScheduleResponse {
  cron: string
  enabled: boolean
  inputs: Record<string, unknown> | null
  timezone: string
  next_run_at: string | null
}

export interface WorkflowScheduleUpdate {
  cron: string
  enabled: boolean
  inputs?: Record<string, unknown> | null
  timezone?: string
}

// --- Test Node types ---

export interface NodeTestRequest {
  node_id: string
  variables: Record<string, unknown>
  env_vars?: Record<string, string>
}

export interface NodeTestResponse {
  node_id: string
  node_type: string
  status: "completed" | "failed"
  output: unknown
  error: string | null
  duration_ms: number
  variables_after: Record<string, unknown>
}

// --- Per-node validation state injected into node data for canvas overlay ---

export interface NodeValidationState {
  errors: string[]
  warnings: string[]
}

// --- Version types ---

export interface WorkflowVersionResponse {
  id: string
  workflow_id: string
  version_number: number
  blueprint: WorkflowBlueprint
  input_schema: Record<string, unknown> | null
  output_schema: Record<string, unknown> | null
  change_summary: string | null
  created_by: string | null
  created_at: string
}
